dxr/static/js/dxr.js

```diff
@@ -322,22 +323,39 @@ $(function() {
         data.query_string = $.param(params);
 
         // If no data is returned, inform the user.
-        if (!data.results.length) {
+        if (!data.results.length && !data.promoted.length) {
             contentContainer
                 .empty()
                 .append(nunjucks.render('partial/results_container.html', data));
         } else {
-            var results = data.results;
-            resultsLineCount = countLines(results);
-
-            for (var result in results) {
-                var icon = results[result].icon;
-                var resultHead = buildResultHead(results[result].path, data.tree, icon, results[result].is_binary);
-                results[result].iconClass = resultHead[0];
-                results[result].pathLine = resultHead[1];
-            }
+            resultsLineCount = countLines(data.results);
+            [data.results, data.promoted].forEach(function(results) {
+                for (var i = 0; i < results.length; i++) {
+                    var icon = results[i].icon;
+                    var resultHead = buildResultHead(results[i].path, data.tree, icon, results[i].is_binary);
+                    // After we build the result head, join up the path so we can set it as an href.
+                    results[i].path = results[i].path.join('/');
+                    results[i].iconClass = resultHead[0];
+                    results[i].pathLine = resultHead[1];
+                }
+            });
 
             if (!append) {
+                // If we have promoted results then pass in a formatted query which bubbles the
```

Comment:
I would far rather do this server-side than try to reimplement query parsing in a half-way fashion here. For example, this will do the wrong thing if a query term has a colon or a space in a quoted string. Let's just construct the "more promoted results" query on the server and pass it down as part of the promoted results data structure.